AVIF support

morsssss opened the pull request:
Demand for AVIF support on the web is growing, as the word gets out about this new file format which allows higher-quality encoding at smaller sizes. Core contributors to major open-source CMSs are interested in auto-generating AVIF images! They've been simply waiting for support to appear in libgd.

This PR aims to meet the growing demand, and to help bring smaller, more beautiful images to more of the web - to sites created by experienced developers and CMS users alike.

This PR adds support by incorporating [libavif](https://github.com/AOMediaCodec/libavif). Another option would be to use [libheif](https://github.com/strukturag/libheif). However, it's generally felt that libavif has more complete support for the AVIF format. libavif is also used by [the Chromium project](https://chromium-review.googlesource.com/c/chromium/src/+/2101317) and [squoosh.app](https://github.com/GoogleChromeLabs/squoosh/pull/804).

In this PR, I've endeavored to incorporate the latest research into best practices for AVIF encoding - not just for default quantizer values, but also an algorithm for determining the number of horizontal tiles, vertical tiles, and threads.

Since this is my first contribution to this project, I've made this a draft PR. A couple of TODOs remain. And during the next days I will be adding more tests.

Thanks for considering this! This fixes #557 .

vapier commented on CMakeLists.txt:
```diff
@@ -119,6 +120,10 @@ else (USE_EXT_GD)
 	ENDIF (ENABLE_WEBP)
 
+	IF (ENABLE_AVIF)
+	  FIND_PACKAGE(AVIF REQUIRED)
```
indent only with tabs

morsssss replied:
got it

vapier commented on configure.ac:
```diff
@@ -278,13 +278,21 @@ GD_LIB_PKG_CHECK([LIBTIFF], [TIFF], [tiff], [libtiff-4], [
 
 dnl Check for webp support.
-GD_LIB_CHECK([LIBWEBP], [WEBP], [webp], [
+GD_LIB_PKG_CHECK([LIBWEBP], [WEBP], [webp], [libwebp], [
```
looks like you should rebase onto the latest state as this has been merged already

`git fetch && git rebase`

vapier commented on cmake/modules/FindAVIF.cmake:
```diff
@@ -0,0 +1,92 @@
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+# TODO: should we remove ../libavif/include path?
```
yes, yes you should

morsssss replied:
true, that!

vapier commented on cmake/modules/FindAVIF.cmake:
```diff
@@ -0,0 +1,92 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
```
this should be checking pkg-config first.  look at how other modules use `PkgConfig`.

cryptomilk replied:
The file `cmake/modules/FindAVIF.cmake` should be removed. libavif ships a cmake config file (something like pkg-config but for cmake!)

morsssss replied:
I'm afraid I'll need guidance from you here, as I'm newer to this build process. I modeled this module after `FindWEBP.cmake`. I thought the build would need to know where to find libavif's header files - and to make sure they existed. Then we can set `AVIF_FOUND` and `HAVE_LIBAVIF`. So, the way it is now mirrors how other resources are discovered.

If I remove this file, I suppose we'll be assuming that libavif exists as long as the developer builds with `-DENABLE_AVIF=1` and `config.h.cmake` contains `#cmakedefine HAVE_LIBAVIF`?

morsssss replied:
@cryptomilk, your next comment gave me a hint. `find_package()` can be used to grab settings from an external project.

cryptomilk replied:
If you have an option to turn on building with AVIF support, then you should require the package:

```
if (ENABLE_AVIF)
    find_package(libavif 0.8.2 CONFIG REQUIRED)
endif()
```

If you want automatic detection you would go with:

```
find_package(libavif 0.8.2 CONFIG)
if (TARGET avif)
    # build avif support
endif()
```

The header file information is part of the avif target ...

vapier replied:
we must always have a knob that allows the user to explicitly enable or disable a feature.  it's OK for the default to test & enable the feature if it's available, but there still has to be a manual control knob.

if you look at the configure script, it does this for all the libs we support.

morsssss replied:
excellent! I'd already implemented the first option above: `if (ENABLE_AVIF)` etc.

Of course, now the build can't find the config file. My libavif build placed it in `libavif/CMakeFiles/Export/lib/cmake/libavif/libavif-config.cmake`. Should I have instructed libavif to put export files in a shared location? If so, how would I do that?

vapier replied:
based on earlier changes, i'm guessing you haven't actually installed libavif, just built it locally.  if you install the library, the cmake file should end up in a path like `/usr/lib64/cmake/` or `/usr/share/cmake/Modules/`.  then cmake can find it.

morsssss replied:
Yep, that's it - the build instructions for libavif I was reading included the flag `-DBUILD_SHARED_LIBS=OFF`. Just switched that out, and I'm rebuilding.

morsssss replied:
Finally realized I had to do a `make install` 😭

vapier commented on cmake/modules/FindAVIF.cmake:
```diff
@@ -0,0 +1,92 @@
+set(AVIF_NAMES ${AVIF_NAMES} avif)
+find_library(AVIF_LIBRARY
+  NAMES ${AVIF_NAMES}
+  PATHS "${PROJECT_SOURCE_DIR}/../deps/lib" /usr/lib64 /usr/lib /usr/local/lib
```
delete all paths that use `PROJECT_SOURCE_DIR`

morsssss replied:
Ok. I got this from [FindWEBP.cmake](https://github.com/libgd/libgd/blob/4690d1ceb6e6d743f61c5d14130d36e837fc4c75/cmake/modules/FindWEBP.cmake#L50). Is it safe to remove there too?

vapier replied:
yeah, it should get trimmed from there

vapier commented on src/config.h.cmake:
```diff
@@ -4,7 +4,7 @@
 
 /* Whether to support gd image formats */
-#cmakedefine01 ENABLE_GD_FORMATS
+#define ENABLE_GD_FORMATS @ENABLE_GD_FORMATS@
```
guessing you didn't mean to make this change

morsssss replied:
yeah, I thought I removed this file from the commit, as it's autogenerated... I wanted to ask why it's included in the repo?

vapier replied:
i think this is the only file that one needs for using cmake, so we make a concession in that regard

vapier commented on src/config.h.cmake:
```diff
@@ -75,9 +78,15 @@
 #cmakedefine HAVE_STDINT_H
 
+/* Define to 1 if you have the <stdlib.h> header file. */
+#cmakedefine HAVE_STDLIB_H
```
not sure why these are popping back up ... we cleaned them out.  rebase onto the latest tree and re-run the bootstrap script to see if these get cleaned up.

vapier commented on src/gd.h:
```diff
@@ -627,6 +627,10 @@ BGD_DECLARE(gdImagePtr) gdImageCreateFromWebp (FILE * inFile);
 BGD_DECLARE(gdImagePtr) gdImageCreateFromWebpPtr (int size, void *data);
 BGD_DECLARE(gdImagePtr) gdImageCreateFromWebpCtx (gdIOCtx * infile);
 
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvif (FILE * inFile);
```
i know we've been inconsistent, but we should fix new code at least -- omit the space after the `*`.  you'll want to check all the new lines you add that do this.

morsssss replied:
Got it. Yeah, I couldn't determine the standard from the existing code. Will change my new stuff to omit that space.

vapier commented on src/gd_webp.c:
```diff
@@ -291,7 +291,7 @@ BGD_DECLARE(void) gdImageWebpEx (gdImagePtr im, FILE * outFile, int quality)
 		return;
 	}
-	_gdImageWebpCtx(im, out, quality);
+	gdImageWebpCtx(im, out, quality);
```
this is incorrect ... it should be using `_gdImageWebpCtx`

morsssss replied:
I figured `gdImageWebpCtx()` was superior because it returns `void`, instead of returning an `int` that gets thrown away. Apparently not?

vapier replied:
`gdImageWebpCtx` is just going to call `_gdImageWebpCtx`, so it's only adding overhead to call `gdImageWebpCtx` ?

it's also about using the internal interfaces than having to go through the exported symbols which require a PLT entry

morsssss replied:
got it!

vapier replied:
looks like these changes are still in here when they shouldn't be

morsssss replied:
Sorry, this one dropped off the map. Reverted.

vapier commented on src/gd_webp.c:
```diff
@@ -315,7 +315,7 @@ BGD_DECLARE(void) gdImageWebp (gdImagePtr im, FILE * outFile)
 		return;
 	}
-	_gdImageWebpCtx(im, out, -1);
+	gdImageWebpCtx(im, out, -1);
```
this is incorrect ... it should be using `_gdImageWebpCtx`

vapier commented on tests/CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ if (BUILD_TEST)
 	include_directories (BEFORE ${GD_INCLUDE_DIR} "${GDTEST_SOURCE_DIR}" "${CMAKE_BINARY_DIR}" "${CMAKE_BINARY_DIR}/tests/gdtest")
 
 	SET(TESTS_DIRS
+	  avif
```
fix the indentation

morsssss replied:
yargh!

vapier commented on tests/Makefile.am:
```diff
@@ -98,6 +99,7 @@ EXTRA_DIST += \
 
 # We don't keep any media files in the top dir ... just generated outputs.
 CLEANFILES += \
+  *.avif \
```
fix the indentation

vapier commented on tests/avif/avif2jpeg.c:
```diff
@@ -0,0 +1,55 @@
+
+#include "gd.h"
+
+/* A short program which converts a .avif file into a .jpg file - 
```
move to the top of the file so it's the first thing people see

vapier commented on tests/avif/avif2jpeg.c:
```diff
@@ -0,0 +1,55 @@
+
+	gdImageJpeg(im, out, 75);
+
+  printf("Wrote oufile %s\n. Success!\n", argv[1]);
```
fix indentation

vapier commented on tests/avif/Makemodule.am:
```diff
@@ -0,0 +1,6 @@
+libgd_test_programs += \
```
same here about adding build-time protection when avif isn't enabled

morsssss replied:
thanks for noticing this

vapier commented on tests/avif/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+LIST(APPEND TESTS_FILES
```
look at `tests/jpeg/CMakeLists.txt` -- this will need `IF(AVIF_FOUND)` protection at least so we don't build the tests when avif isn't available

vapier commented on tests/avif/avif2jpeg.c:
```diff
@@ -0,0 +1,55 @@
+	gdImagePtr im;
+	FILE *in, *out;
+
+	if (argc != 3) {
```
i guess this is more of a human util rather than an actual unittest.  these would prob be better under examples/.

but we'll still want an unittest for these APIs.  we'd want a small avif file to check into the tree with a reference png, and then use things like `gdTestImageCompareToFile` to make sure we can read it correctly.  and vice-versa.

that way we can run `make check` and have everything validate w/out user interaction.

morsssss replied:
Makes sense. I have to write some more unit tests for this too...

morsssss replied:
Now also writing the three tests I've seen for every other image format...

morsssss replied:
im2im fails, even though the transformed images look the same to me when I write them to files... there are small changes, like red 0xFF => red 0xFE. Hmm. AVIF is lossy, so I'm not sure I'll get pixel perfection. I've noticed the image comparison is commented out in jpeg_im2im.c.

I'm going to try the approach you mention, which looks like what I see in jpeg/bug_github_18.c.

morsssss replied:
My new tests are failing for the same reason as my old tests - R, G, or B values are often off by one. This even though I'm encoding at a quality that, according to `avif.h`, should be lossless:

`#define AVIF_QUANTIZER_LOSSLESS 0`

I wonder if @wantehchang might recognize what's going on here. Is there a missing bit of precision?

Some samples. (We're comparing `r1` to `r2`, `g1` to `g2`, etc.)

![Screen Shot 2021-02-15 at 6 02 43 PM](https://user-images.githubusercontent.com/29419498/108001163-b8809700-6fb9-11eb-8787-00fc8f400fe8.png)
![Screen Shot 2021-02-15 at 6 05 02 PM](https://user-images.githubusercontent.com/29419498/108001171-bcacb480-6fb9-11eb-903d-a68a4477e049.png)
![Screen Shot 2021-02-15 at 6 05 29 PM](https://user-images.githubusercontent.com/29419498/108001177-bfa7a500-6fb9-11eb-8458-c97c9f97cff2.png)

morsssss replied:
I don't suppose libavif ever loses a bit of precision when convert between YUV and RGB? I don't spot anything obvious in, say, `avifImageYUV8ToRGB8Color`, but a lot of conversion does go on there...

morsssss replied:
One of the tests where I see this is [compare_avif_to_png.c](https://github.com/libgd/libgd/blob/6851185549a4f6c91b8dad5d0d1fa3bfb48ef887/tests/avif/compare_avif_to_png.c).

morsssss replied:
Fixed it. Tests pass. Thanks @wantehchang for the tip about `AVIF_MATRIX_COEFFICIENTS_IDENTITY`.

vapier commented on windows/Makefiletest.vc:
```diff
@@ -1,5 +1,6 @@
 TESTS=bmp\bmp_im2im \
-bmp\bmp_null \
```
shouldn't this keep bmp_null ?

morsssss replied:
Yes. I don't know why I lost that line.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+#include "gdhelpers.h"
+
+#ifdef HAVE_LIBAVIF
+#include "avif.h"
```
this needs to be `<avif/avif.h>` instead

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+#ifdef HAVE_LIBAVIF
+#include "avif.h"
+#include "internal.h"
+// internal.h is only used for AVIF_MIN() and AVIF_CLAMP() macros
```
internal.h isn't installed so you cannot use it.  only avif.h is installed.

morsssss replied:
thanks for noticing that. I see it now...

(I was a little nervous about using a file called `internal.h` anyway)

Is it ok if I reproduce these macros? I see them defined in gd_interpolation.c, but not elsewhere in libgd.

vapier replied:
ugh, what a mess.  rebase onto latest master as i've cleaned things up a little.  if you include gd_intern.h, you'll get MIN/MAX/CLAMP macros.

morsssss replied:
Ok. I'll re-rebase before I keep working on this.

morsssss replied:
Now I see! You just added these macros to gd_intern.h.

I've rebased and I'll happily use your new, globally available, macros.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+#define DEFAULT_QUALITY -1
+#define DEFAULT_SPEED AVIF_SPEED_DEFAULT
+
+// This initial size for the gdIOCtx is standard among GD image conversion functions
```
comments should be full sentences, so put periods at the end of them.  should double check all comments so i don't have to flag them.

morsssss replied:
done!

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+
+/*** Helper function signatures ***/
+
+static avifBool _gdImageAvifCtx (gdImagePtr im, gdIOCtx * outfile, int quality, int speed);
```
omit spaces before `(` with funcs

and after `*` with pointers

morsssss replied:
Many functions in the codebase have a space before the `(` and args... but I assume that you know that and want contributors to follow this convention going forward :)

spaces before left parens in function declarations removed.

morsssss replied:
(and I think I got all the spaces after *'s for pointers yesterday; when I check this in they should be gone)

vapier replied:
yes, i noted somewhere else that we're inconsistent between the two forms.  so i'd like to move in this direction with new code.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvif (FILE * infile)
+{
+  gdImagePtr image;
```
we seem to have settled on `im` instead of `image` in the codebase

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  }
+
+  image = gdImageCreateFromAvifCtx(ctx);
+  // ctx->gd_free(ctx);
```
clean up dead comments

although i think you wanted to call this

morsssss replied:
I did want to call this. I commented it out because it gave me an error:

`malloc: *** error for object 0x100d060f0: pointer being freed was not allocated\r\n"`

even though this memory had in fact been allocated by `gdNewFileCtx()`.

Looking around the code, I even see the exact same construction in `gdImageCreateFromJpegEx()`. 

Any idea what I might be missing here?

vapier replied:
that looks weird.  kind of feels like it might be memory corruption ?

you could try building libavif & gd with sanitizers enabled and run the unittests and see if they catch anything interesting.

morsssss replied:
I haven't done this... but I poked around some more without finding any ideas.

I'd feel bad not releasing this memory, but it's very few bytes....

morsssss replied:
I figured it out. I was already freeing this up in the `destroyCtxAndAvifIO` callback.

That wasn't a good idea. Now that's just `destroyAvifIO`, and we destroy the ctx in the function where it was created.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvif (FILE * infile)
+{
+  gdImagePtr image;
+  gdIOCtx *ctx = gdNewFileCtx(infile);
```
only indent with tabs

morsssss replied:
Just converted all my space indentations in this file to tabs

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  gdImagePtr image;
+  gdIOCtx *ctx = gdNewFileCtx(infile);
+
+  if (!ctx) {
```
you can omit braces when the body is only one statement

morsssss replied:
You don't know how glad I am to hear you say that :)

vapier replied:
that's not really an issue.  compilers have `-Wmisleading-indentation` nowadays and they would have caught that OpenSSL bug.

morsssss replied:
I'll wait for a consensus here...

vapier replied:
you can omit braces ;)

morsssss replied:
Wrote a regex and got rid of 17 instances of these braces.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  gdIOCtx *ctx = gdNewFileCtx(infile);
+
+  if (!ctx) {
+    return GD_FALSE;
```
the return value is a pointer, not a bool, so this needs to be NULL

morsssss replied:
thanks for catching that. Just also added error handling in the function that calls this one.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  avifDecoder *decoder;
+  avifRGBImage rgb;
+
+  decoder = avifDecoderCreate();
```
do any of these avif calls need their own error checking ?  i.e. if any of them fail, you need to check & return errors accordingly.

morsssss replied:
Some do! This one doesn't; if it can't `malloc`, it `abort`s.

Checked the rest of my avif calls, and for the ones that can return an error code, I've already got a check. The others don't return anything.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  return _noAvifError();
+}
+
+#endif /* HAVE_LIBAVIF */
```
add missing newline to end of file

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+static void setEncoderTilesAndThreads(avifEncoder * encoder, avifRGBImage * rgb);
+
+
+/*****************************************************************************
```
this file isn't big, so you can omit these giant comment blocks

morsssss replied:
removed

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+
+BGD_DECLARE(void) gdImageAvif (gdImagePtr im, FILE * outFile)
+{
+	return gdImageAvifEx(im, outFile, DEFAULT_QUALITY, AVIF_SPEED_DEFAULT);
```
this is a void func, so return doesn't make sense

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  }
+
+  result = avifImageRGBToYUV(avifIm, &rgb);
+  if ((failed = isAvifError(result, "Could not convert image to YUV"))) {
```
don't inline assignments into `if` statements

morsssss replied:
Understood. It's so convenient. But not everyone likes it. (The compiler didn't until I added the extra layer of parens)

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+#define MAX_THREADS 64
+
+
+/*** Helper function signatures ***/
```
i'd prefer to move these small helper funcs up here if possible and avoid having to duplicate the decls

morsssss replied:
done

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  int imageArea, tiles, tilesLog2;
+  int tileRowsLog2, tileColumnsLog2, maxThreads;
+
+  imageArea = rgb->width * rgb->height;
```
use `overflow2` before doing any multiplication

morsssss replied:
Thanks. I saw that elsewhere in the code, but didn't implement it. Now I did. Integer overflow can be dangerous:

https://www.nytimes.com/2015/05/01/business/faa-orders-fix-for-possible-power-loss-in-boeing-787.html

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+   If so, decode the error and output it, and return true.
+   Otherwise, return false.
+*/
+static avifBool isAvifError(avifResult result, char * msg) {
```
msg should be `const`

morsssss replied:
I thought I'd change this to `const char* const`. But I see that gd almost always just uses `const char *`, including the parallel situation of `gd_error()`. So I've gone with the standard... let me know if I should specify the double immutability instead!

vapier replied:
C/C++ rules around `const` are annoying.  using `const char *` is common and is largely sufficient as it means the memory it points to is constant (which is what we want).  using `const char * const` makes the pointer itself constant which isn't that useful.  it's like how `const int` and `const bool` are rarely used since they're passed by value, not by reference.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+static avifBool isAvifError(avifResult result, char * msg) {
+  if (result != AVIF_RESULT_OK) {
+    gd_error("avif error: %s: %s", msg, avifResultToString(result));
+    return GD_TRUE;
```
the return is `avifBool`, but you're using gd's `GD_TRUE` ... should be consistent

morsssss replied:
True. I ended up changing all my GD_(TRUE|FALSE) to AVIF_(TRUE|FALSE), since there's no GD "boolean" type, and since the other instances were placed into libavif structures.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+static avifIO *createAvifIOFromCtx(gdIOCtx * ctx) {
+  avifIO *io;
+
+  io = (avifIO *) gdMalloc(sizeof(avifIO));
```
`gdMalloc` return is `void*`, and C doesn't require casting like C++, so just omit it

use `sizeof(*io)` instead

morsssss replied:
the rest of the code is full of examples of casting the result of `gdMalloc()` - but I assume that you want to remove this unnecessary practice going forward. So I'll omit it!

Similarly I'll change `sizeof()` to refer to the variable instead of the type.

morsssss replied:
(though I'm curious why you prefer that `sizeof()` refer to the variable itself)

vapier replied:
i've been bitten by more hardcoded size values to malloc not being updated than the pointer changing types, so i prefer that style

morsssss replied:
that makes sense!

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  }
+
+  // Try to seek offset bytes forward. If we pass the end of the buffer, throw an error.
+  if (!(ctx->seek(ctx, offset))) {
```
inner parens are unnecessary

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  // If getBuf() returns a negative value, that means there was an error.
+  int charsRead = ctx->getBuf(ctx, dataBuf, size);
+  if (charsRead < 0) {
+    return AVIF_RESULT_IO_ERROR;
```
double check your error code paths so you don't leak (e.g. the `gdRealloc` above leaks)

morsssss replied:
Thanks, fixed that. I just hope the AVIF functions free up this buffer eventually. I haven't checked that, but I don't have much control over it.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+    return AVIF_RESULT_IO_ERROR;
+  }
+
+  dataBuf = gdRealloc(dataBuf, size);
```
while not incorrect, why realloc instead of malloc ?  databuf is always NULL here.

morsssss replied:
True. This is an artifact of an earlier approach. Changing to malloc.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+
+  //TODO: if we set sizeHint, this will be more efficient.
+
+  if (offset > LONG_MAX || size < 0) {
```
`size` is a `size_t` so, by definition, can never be negative, so comparing it to `< 0` will always be false

morsssss replied:
Removing it. What you say is true. I added this because I saw something similar [in libavif](https://github.com/AOMediaCodec/libavif/blob/be4a2139d7dfe44c77ca38e678fbdf30b2471349/src/io.c#L101) - but I suppose that was there to ensure that `size` was nonzero, not <0, which it can't be.

I figured I'd allow `size` to be 0. Maybe I shouldn't.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+
+// avif.h says this is optional, but it seemed easy to implement.
+static void destroyCtxAndAvifIO(struct avifIO *io) {
+  gdIOCtx *ctx = io->data; // TODO: should I have cast this?
```
no, because this is C, and assigning void* to any other pointer type is fine

cryptomilk commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+// Read the pixels from the AVIF image and copy them into the GD image.
+// Image depth can be 8, 10, 12, or 16. But if depth>8, pixels are 16-bit.
+
+  if (rgb.depth == 8) {
```
Could you please use a switch statement here:

```
swtich (rgb.depth) {
    case 12:
    case 10:
        /* 16 bit conversion */
    case 8:
       /* 8bit conversion */
    default:
        /* error our */
}
```

morsssss replied:
I felt like the comment covered the reasoning here. But you'd prefer if the reasoning was embedded in the code? Or are you more concerned about cases where the image depth would be something other than 8, 10, 12, or 16?

Also, doesn't libavif allow for 16-bit pixel depth?

`uint32_t depth;       // legal depths [8, 10, 12, 16]. if depth>8, pixels must be uint16_t internally`

cryptomilk replied:
There is no 16bit as it would just be a waste of space.

morsssss replied:
True, I've never seen anything saying AVIF will support 16-bit pixels - just 8, 10, and 12. And yet libavif explicitly says that 16 bits are legal... and, as far as I can tell, 10-bit and 12-bit pixels get stored in 16-bit sizes anyway.

I guess, my point is, if libavif allows for 16-bit pixels, is there harm in gd supporting that?

If you think so, I'll drop it.

vapier replied:
seems like including 16 costs nothing to gd, so if libavif supports it, not a big deal to leave it in

cryptomilk replied:
AVIF doesn't support 16bit. We are decoding an AVIF image here and if it says it is 16bit there is probably something wrong. If the bit depth would be set to 0xdeadbeef you would still decode it ...

wantehchang replied:
Here we know the avifDecoderNextImage() call succeeded, so decoder->image->depth can only be 8, 10, or 12 (the three bit depths supported by AV1). We also know rgb.depth was set by the avifRGBImageSetDefaults() call to be equal to decoder->image->depth. So rgb.depth can only be 8, 10, or 12.

Note: Although avifRGBImageSetDefaults() can be modified to use different defaults, it is extremely unlikely that the default for rgb.depth will change.

morsssss replied:
Ok! I've implemented the `switch` construction and error message that @cryptomilk suggests.

cryptomilk commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+    }
+  }
+
+  /* do not use gdFree here, in case gdFree/alloc is mapped to something else than libc */
```
I completely miss color profile handling here, ICC or NCLX boxes ...

morsssss replied:
Reading your comment, I looked around the libgd codebase, and I didn't see explicit support for color profiles anywhere else - not even in PNGs, where I believe this is included in the spec.

Of course, libavif supports embedded ICC profiles explicitly.

But since it's not something gd deals with, I'm not sure if it would make sense to add it here. How do you think this should be handled?

vapier replied:
i don't think we want to hold up adding libavif support for color profiles if we aren't doing color profiles anywhere yet

cryptomilk replied:
I would say this is a big problem. AVIF has been created to be able to handle HDR images e.g. P3 PQ or HG. It means we can only handle sRGB files.

So we have to check the NCLX box and if it isn't sRGB stop and if it has an ICC profile we have to abort too.

Here are test files:
https://github.com/AOMediaCodec/av1-avif/tree/master/testFiles/Netflix/avif

vapier replied:
i don't think we're disagreeing that we should have color profile support in general, but we are disagreeing that it needs to be implemented *before* merging libavif support.  feel free to comment on #136.

morsssss replied:
Do you think we need to check the icc field when decoding an AVIF image - and abort if one is found? Or can we just leave this for now? I don't see a way that the NCLX status is stored explicitly.

If it helps, I believe the primary use case here will be for web images. Though @vapier would know more.

cryptomilk replied:
I did not say you should add support for color profiles here, however if the AVIF file isn't using sRGB which is the web standard it doesn't make sense to convert to any other image format as the result will look just horrible on your screen.

So when coding check if the NCLX box says it is sRGB we fine. If it isn't sRGB you will end up with garbage (horribly looking image).

morsssss replied:
ok - sounds like I should make sure the color profile is sRGB, and, if it's not, throw an error.

morsssss replied:
I looked at this for a little while tonight. Trying decoding on an image that doesn't have ICC info, I saw this:

![Screen Shot 2021-02-10 at 8 42 55 PM](https://user-images.githubusercontent.com/29419498/107595325-54895780-6be2-11eb-92b9-5b2705c7f1ee.png)

I could certainly throw an error if values were set here that did *not* correspond to sRGB. If I got sRGB values, or I got the `UNSPECIFIED` values in the screenshot, I wouldn't throw an error.

To see how this struct looks for sRGB images, I looked at the test files you sent. As far as I can tell, the AVIF images were encoded without ICC:

`avifenc -c aom --speed 0 --jobs 12 --depth 8 --ignore-icc ...`

So I downloaded an HDR PNG and encoded that to AVIF, then ran it though my decoder without that `--ignore-icc` flag. Sure enough, I saw other values:

![SDR image](https://user-images.githubusercontent.com/29419498/107599669-4988f400-6bef-11eb-9f53-ed5bda100333.png)

Presumably, if I saw this, I'd throw an error.

I then downloaded an SDR PNG from the same repo and encoded to AVIF, and ran it through the decoder... but this time I got the same values as I did from the HDR PNG. Hmm.

Am I doing this correctly? Do you have an AVIF image that's got an sRGB ICC profile attached that you could send me for testing?

cryptomilk replied:
If you convert a png to an avif and an ICC profile is attached to the png, you should just copy the icc blob over to the avif file. You don't know its content. If there is no ICC you should just assume it is sRGB and write a sRGB NCLX box. If you have an AVIF with an ICC profile, convert it and write the ICC profile to the image ...

An NCLX box holds the same information as an ICC profile. However we are not able to convert an NCLX box to an ICC profile or vice versa. At least right now. You would need littlecms2 for this.

If you decode an AVIF file and it has an NCLX box which is not sRGB we need to stop. It doesn't make sense to convert the image as the resulting file can be correctly displayed by any software.

paperboyo replied:
I’m just a user, so don’t know much, please treat this comment with a truckload of salt. As noted, the library is currently fully colour blind. Regarding stopping processing on anything non-sRGB, this is quite drastic and would make it unpredictable in any automated scenario (some files would just bomb). AVIF aren’t any different to other files whose profiles library currently is blind to, if maybe for the fact that AVIF is more likely to have exotic, wide gamut data. The further source data is from sRGB, the more insane is the output, but, again, it also applies to any other format (eg. a PNG with ProPhotoRGB converted to a PNG).

>If you convert a png to an avif and an ICC profile is attached to the png, you should just copy the icc blob over to the avif file.

This is scenario no. 1 [from here](https://github.com/libgd/libgd/issues/136#issuecomment-318921312). While not exactly correct, it’s less incorrect than current situation (dropping the profile). Would be great if this behaviour could apply to all images, not just AVIF. “Great”, that is, until the library becomes fully colour-aware at some point in the future.

morsssss replied:
"I'm just a user" - that means you're a vital part of this conversation! Thanks for putting this here.

I've implemented @cryptomilk's suggestion to detect non-sRGB images and throw an error, but I'm honestly more comfortable with what you suggest. I feel that someone who decides to convert a non-sRGB image shouldn't be surprised if the result isn't great.

I see that `gd_errors.h` allows for warnings, not just fatal errors - and that files like `gd_jpeg.c` and `gd_png.c` use warnings. Perhaps that a warning would be a good compromise? Even though I don't think warnings would be visible to people who use GD via, say, the PHP wrapper.

morsssss replied:
Changed this to emit a warning.

cryptomilk reviewed (changes requested):
I've added some comments here. For now I just looked at decoding. I think we should get that right before we move on to encoding which is more tricky ...

vapier commented on CMakeLists.txt:
```diff
@@ -119,6 +120,10 @@ else (USE_EXT_GD)
 	ENDIF (ENABLE_WEBP)
 
+	IF (ENABLE_AVIF)
+	  FIND_PACKAGE(AVIF REQUIRED)
```
if we set a min ver, we should do it in configure.ac too

i don't have an opinion on which ver to use beyond "whatever API version this patch is compatible with"

morsssss commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  }
+
+  // Set up the avifRGBImage. Use default settings unless there's demand to override these.
+  avifRGBImageSetDefaults(&rgb, decoder->image);
```
`avifRGBImageSetDefaults()` sets `rgb.format = AVIF_RGB_FORMAT_RGBA;`. 

What am I missing here?

morsssss replied:
let me know if I'm still missing something here! Otherwise I can resolve this one...

wantehchang replied:
Perhaps cryptomilk would prefer to see rgb.format set to AVIF_RGB_FORMAT_RGBA explicitly?

morsssss replied:
As per @wantehchang 's suggestion in chat, I've instead proposed to document this default in libavif:

https://github.com/AOMediaCodec/libavif/pull/492

morsssss commented on CMakeLists.txt:
```diff
@@ -171,6 +176,11 @@ else (USE_EXT_GD)
 	ENDIF(WEBP_FOUND)
 
+	IF(AVIF_FOUND)
+		INCLUDE_DIRECTORIES(${AVIF_INCLUDE_DIR})
```
ok! let me try this and make sure I can still build.

morsssss replied:
@cryptomilk, I've had trouble getting this to work. I think we need your superior experience with CMake here!

Looking into this deeply - `libavif-config.cmake` contains the line `add_library(avif SHARED IMPORTED)`. That seems promising. But I don't know the mechanism by which that library gets added. 

In my current build process, `find_library()` locates `libavif` and `aom` and adds the relevant info to `AVIF_LIBRARIES`. These then propagate to `LIBGD_DEP_LIBS`, which get added to the list of things to link by `target_link_libraries(${GD_LIB} ${LIBGD_DEP_LIBS})`.

I'm not sure how to replicate that process with `libavif-config.cmake`.

Would you consider fixing this yourself in a followup PR?

cryptomilk replied:
Did you use `find_package(libavif 0.8.2 CONFIG)` and remove the `cmake/Modules/FindAVIF.cmake` file?

All you have to do is:

```
if (TARGET avif)
  append(LIBGD_DEP_LIBS avif)
endif()
```

morsssss replied:
I did indeed use `find_package(libavif 0.8.2 CONFIG REQUIRED)` and remove `cmake/Modules/FindAVIF.cmake`!

I can try this - but what about `aom`? The build needs to find and link that library too.

cryptomilk replied:
Why would it need to link aom?

morsssss replied:
I'd included that - but maybe it's unnecessary.

I just tried your suggestion - thanks for that! I had to amend it, though:

```
if (TARGET avif)
	LIST(APPEND LIBGD_DEP_LIBS avif)
endif()
```

Unfortunately this just appended the word `avif` to the list of libraries:

`.../usr/local/lib/libwebp.a;avif`

I think it would save us both several more hours if you were willing to make this change via a followup PR :)

cryptomilk replied:
This looks correct, `avif` is a cmake target!

morsssss replied:
Sadly, it just appended the word `avif` to the list of libraries. It didn't work for me 😭

morsssss replied:
Thanks for submitting that patch. That... does work for me!

😅

vapier commented on src/gd_filename.c:
```diff
@@ -37,7 +37,7 @@ static void BGD_STDCALL writebmp(gdImagePtr im, FILE *out) {
 
 
-enum FType {UNKNOWN, PNG, JPG, GIF, TIFF, GD, GD2, WEBP};
+enum FType {UNKNOWN, PNG, JPG, GIF, TIFF, GD, GD2, WEBP, AVIF};
```
please rebase onto the latest master ... i've deleted this enum as it's unused

morsssss replied:
Hmm, I thought I'd done this, but it doesn't look like I did it right. I'll re-rebase

cryptomilk commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+      *(p++) = a;
+    }
+  }
+
```
You need to set the yuvRange and fill out the nclx box before you convert from RGB to YUV:

```
avifIm->yuvRange = AVIF_RANGE_FULL;

/ Set to sRGB per default as color profiles are not supported
avifIm->colorPrimaries = AVIF_COLOR_PRIMARIES_BT709;
avifIm->transferCharacteristics = AVIF_TRANSFER_CHARACTERISTICS_SRGB;
avifIm->matrixCoefficients = AVIF_MATRIX_COEFFICIENTS_BT709;
```

morsssss replied:
ok, thanks!

cryptomilk commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+		return 1;
+	}
+
+  avifImage *avifIm = avifImageCreate(gdImageSX(im), gdImageSY(im), 8, DEFAULT_CHROMA_SUBSAMPLING);
```
I would use AVIF_PIXEL_FORMAT_YUV422 above 90% quality.

morsssss replied:
That's an interesting idea...

@wantehchang, do you have an opinion here? The idea intrigues me... my only concern is that it may make our `quality` param more obtuse than it already is.

wantehchang replied:
The YUV 4:2:2 format is not commonly used. You can use AVIF_PIXEL_FORMAT_YUV444 if you don't want chroma subsampling when the quality level is above 90%.

morsssss replied:
Did this!

cryptomilk commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+
+  imageArea = rgb->width * rgb->height;
+
+  tiles = ceil(imageArea / MIN_TILE_AREA);
```
Using tiles makes encoding faster, but at the same time the file size increases. So what is a goal of gd here, making encoding fast or produce smaller images?

I guess we want smaller files ...

You should use a tile size of 2048x2048. If the dimension of the image get bigger the additional size for tiled images is negligible. If the image with or height is getting above 4096 use increase the tile size to the max of 3072.

vapier replied:
gd defaults would be commonly accepted defaults for the particular library which balance the two.  then the `Ex` variants of the APIs allows the user to tweak the knobs.

Ben & i talked about exposing lower library-specific state so users could tweak things even more, but we were going to wait on that until after this lands as it's an issue that applies to all backends (png/webp/jpeg/etc...).

morsssss replied:
Yeah, that's a good question. My concern is that AVIF encoding can get really slow. But I see your point too.

The algorithm here is based on research done by a number of people. @wantehchang, do you want to share your experience here?

morsssss replied:
Brief anecdotal, non-scientific test: I just encoded a 4000x3000 image with the default speed and with quantizer=30. I tried it with 4 tiles, 8 tiles, and then 32 tiles. The number of bytes in the resulting file barely varied - 446993, 447012, and 447210. The time varied more - 52 seconds, 44 seconds, and 38.

I'll keep trying this with other params.

wantehchang replied:
I first want to clarify that the tiles we are discussing here are the tiles in AV1. The MIAF standard, which the AVIF image format is based on, supports image grids and also refers to the elements of an image grid as tiles. (libavif calls them "grid cells".) Those are not the tiles we are discussing here.

Regarding the use of tiles in AV1, I can only provide two general guidelines from my colleagues who work on the libaom encoder.

> Avoid small tiles because they are particularly bad for image coding. Tiles improve encoding and decoding speeds when multiple threads are available. However, for image coding, the total tile boundary length affects the compression efficiency because intra prediction can't go across tile boundaries. So the more tiles there are in an image, the worse the compression ratio is. For a given number of tiles, making the tile size close to a square tends to reduce the total tile boundary length inside the image.

> Use no more tiles than the number of threads. Aim for one tile per thread. Using more than one thread inside one tile could be less efficient. Using more tiles than the number of threads would result in a compression penalty without much benefit.

Tiles can help speed up decoding because tiles can be decoded in parallel. So that's another factor to consider when deciding whether tiles should be used.

The 512x512 minimum tile area was suggested by a colleague of mine. You can certainly use a larger minimum, but 2048x2048 seems too big. Alternatively, you can deal with tiles in a follow-up pull request.

morsssss replied:
If no one objects, I'll stick with what we have now - which is based on current research by a number of people. Let me know if we should revisit this.

wantehchang commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+// For computing the number of tiles and threads to use during encoding
+// Maximum threads are from libavif/contrib/gkd-pixbuf/loader.c
+#define MIN_TILE_AREA (512 * 512)
+#define MAX_TILES 6
```
MAX_TILES should be a power of 2 because the number of tile rows and the number of tile columns must be powers of 2.

morsssss replied:
Yes! I'd noticed that one and changed it to 8 last night. Looks like the new code isn't showing up in this view :)

wantehchang replied:
Ben: I am sorry I commented on old issues that have been fixed. (I responded to your questions for me, which were posted on an old view of the PR, and naturally looked at related code.)

morsssss replied:
normally I wouldn't manage to spot these two issues so quickly :)

wantehchang commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  tiles = AVIF_MIN(tiles, MAX_TILES);
+  tiles = AVIF_MIN(tiles, MAX_THREADS);
+
+  tilesLog2 = floor(log10(tiles) / log10(2));
```
C99 has the log2() function: https://en.cppreference.com/w/c/numeric/math/log2

morsssss replied:
Thank you!

morsssss replied:
So civilized...

wantehchang commented on src/gd_avif.c:
```diff
@@ -0,0 +1,667 @@
+  }
+
+  // It's good to have one thread per tile.
+  encoder->maxThreads = tiles;
```
This should be set to the final number of tiles, based on encoder->tileColsLog2 and encoder->tileRowsLog2:

  encoder->maxThreads = (1 << encoder->tileRowsLog2) * (1 << encoder->tileColsLog2);

morsssss replied:
Something else I noticed last night and changed, but which apparently is not reflected in this view :) This line now reads:

```
encoderTiles = (int) (pow(2, encoder->tileRowsLog2) * pow(2, encoder->tileColsLog2));
encoder->maxThreads = encoderTiles;
```

your bit-shifting approach looks more efficient.

thanks for checking all of this so carefully!

cryptomilk commented on src/gd_avif.c:
```diff
@@ -0,0 +1,646 @@
+/*
+	 As of February 2021, this algorithm reflects the latest research on how many tiles
+	 and threads to include for a given image size.
+	 This is subject to change as research continues.
```
This algorithms could be much simpler. There is no need to calculate tiles first. Also the floor(log2()) can just be a small lookup table.

      encoder->tileColsLog2 = floor_log2(width / width_tile_size) / 2;
      encoder->tileRowsLog2 = floor_log2(height / height_tile_size) / 2;

vapier replied:
lets avoid premature optimization with building lookup tables.  i suspect 2 x (floor+log2) will be system noise compared to the AV1 codec in libavif itself.

lets merge a working & straight forward implementation, and if you want to profile the code to find hot spots, we can do it as a follow up.  but i strongly suspect GD is never the hot spot.

morsssss replied:
Agreeing with that - this operation is only done once per encoding, and it will take a fraction of a second. We'll spend much more time in the AVIF encoding process, which can easily take 1-60 seconds.

morsssss replied:
Also, if it helps, I'm changing this to the version @wantehchang proposed, using `log2()` directly.

wantehchang replied:
FloorLog2 is a common operation in this kind of code. libgd may already have an optimized implementation that goes by some other name such as "count leading zeros" or "bit scan reverse". See https://en.wikipedia.org/wiki/Binary_logarithm#Integer_rounding.

morsssss commented on src/gd_avif.c:
```diff
@@ -0,0 +1,646 @@
+		goto cleanup;		
+	}
+
+	// memset(rgb, 0, sizeof(rgb));
```
sorry, I'd removed that but hadn't committed since then. Just did so!

wantehchang commented on src/gd_avif.c:
```diff
@@ -0,0 +1,646 @@
+				uint8_t g = *(p8++);
+				uint8_t b = *(p8++);
+				uint8_t a = *(p8++);
+				im->tpixels[y][x] = gdTrueColorAlpha(r, g, b, a);
```
Ben: If you want the minimum changes, I suggest the following:

1. Set rgb.depth = 8 between the avifRGBImageSetDefaults() and avifRGBImageAllocatePixels() calls. You can change the comment at line 361 to say "8-bit RGB".
2. Update the comment at lines 376-377, or just delete those two lines.
3. Delete line 379 ( "if (rgb.depth == 8) {" ).
4. In line 386, pass a >> 1 as the last argument to gdTrueColorAlpha().
5. Delete the entire else block (lines 390 - 400).

morsssss replied:
Thanks for catching this. It was even worse than that. Not only do I have to shift the alpha by one bit to convert between 8-bit and 7-bit, but I have to reverse it by subtracting it from 0xFF.

morsssss replied:
I also wasn't daring enough to try to try out the solution you mentioned that would use the bytes as we get them, setting BGRA/ARGB according to endianness - even though I liked the idea of saving the time used to copy everything over.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,663 @@
+// Since avifDecoderNextImage() has succeeded, we know we've got a valid pixel depth.
+
+	if (rgb.depth == 8) {
+		for (y = 0, p8 = rgb.pixels; y < decoder->image->height; y++) {
```
i guess you copied it from gd_webp, but move the `p8` assignment out of the for loop init.  it's unnecessary and doesn't improve readability.

morsssss replied:
yeah, agreed!

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,663 @@
+		}
+		
+	} else {
+		for (y = 0, p16 = (uint16_t *) rgb.pixels; y < decoder->image->height; y++) {
```
i guess you copied it from gd_webp, but move the `p16` assignment out of the for loop init.  it's unnecessary and doesn't improve readability.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,663 @@
+{
+	int x, y;
+	uint8_t *p8; 
+	uint16_t *p16;
```
these pointers are only used in the limited scope below, so prefer to move them down to where they're actually used

morsssss commented on src/gd_avif.c:
```diff
@@ -0,0 +1,646 @@
+				uint16_t g = *(p16++);
+				uint16_t b = *(p16++);
+				uint16_t a = *(p16++);
+				im->tpixels[y][x] = gdTrueColorAlpha(r, g, b, a);
```
I just took your second suggestion. I had borrowed this from an AVIF demo program, but it was useless here. thanks!

morsssss commented:
Looks like we're close here! I'd like to resolve outstanding issues and see if this is ready for merge. Here's what I see:

### CMake build process
The current process works. We'd like to get rid of `FindAVIF.cmake` in favor of the process that comes with libavif. Unfortunately, I think only @cryptomilk has the background to make this work. Could this be done in a followup PR?

### Color profiles
When the user requests that we decode an AVIF image that's not sRGB, should we throw an error, or simply let them do it? I'd be inclined to simply emit a warning, but I don't know if developers using a language wrapper would see this.

### Examples
@vapier , I created the human-usable encoding and decoding programs in `tests/` because we knew they would build there easily :) Do you think I should:
* delete those
* add them to `examples/`
* add them to `GD_PROGRAMS`

### Tests
im2im tests are currently failing because R, G, and B values are often off by 1. I'm not sure why this is happening - if the lossless encoding isn't truly lossless, whether a bit of precision is being lost somewhere, or I've done something I shouldn't have.

Should I add some more basic unit tests in any case?

### Etc
Is there anything else we need here?

Thanks, everyone!

morsssss commented:
Update: lossless conversion is now 100% lossless, and all current tests pass.

cryptomilk commented:
You need to rebase your branch on the master branch of libgd.

Here is a patch which applies to your current tree:
https://github.com/cryptomilk/libgd/commit/0b6804c903753c30cb7e3399dfdd2e64055aa7d4

```
ldd Bin/libgd.so
        linux-vdso.so.1 (0x00007ffcab743000)
        libz.so.1 => /usr/lib64/libz.so.1 (0x00007f20a60e2000)
        libpng16.so.16 => /usr/lib64/libpng16.so.16 (0x00007f20a6099000)
        libavif.so.9 => /usr/lib64/libavif.so.9 (0x00007f20a607d000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f20a5eae000)
        libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f20a5c95000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f20a5b50000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f20a5b33000)
        libdav1d.so.5 => /usr/lib64/libdav1d.so.5 (0x00007f20a59f9000)
        librav1e.so.0 => /usr/lib64/librav1e.so.0 (0x00007f20a575a000)
        libaom.so.2 => /usr/lib64/libaom.so.2 (0x00007f20a5283000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f20a6189000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f20a527c000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f20a525b000)
```

morsssss commented:
_Update 2 - cryptomilk has fixed the cmake build process. Here's what remains:_

### Color profiles
When the user requests that we decode an AVIF image that's not sRGB, should we throw an error, or simply let them do it? I'd be inclined to simply emit a warning, but I don't know if developers using a language wrapper would see this.

### Examples
@vapier , I created the human-usable encoding and decoding programs in `tests/` because we knew they would build there easily :) Do you think I should:
* delete those
* add them to `examples/`
* add them to `GD_PROGRAMS`

### Tests
* Should I add some more basic unit tests?
* A Travis test is failing, I believe because config.h.cmake is different?

vapier commented on CMakeLists.txt:
```diff
@@ -119,6 +120,11 @@ else (USE_EXT_GD)
 	ENDIF (ENABLE_WEBP)
 
+	IF (ENABLE_AVIF)
+	  FIND_PACKAGE(libavif 0.8.2 REQUIRED CONFIG)
```
only use tabs to indent

morsssss replied:
sorry, I keep grepping for spaces and fixing them... and then I make a change and don't do the grep :-/

vapier commented on src/CMakeLists.txt:
```diff
@@ -134,11 +135,18 @@ SET(LIBGD_DEP_LIBS
 	${RAQM_LIBRARIES}
 )
+
+if (TARGET avif)
```
you shouldn't need this `if` block at all.  just adding `${AVIF_LIBRARIES}` above should be sufficient.  if it isn't, it means `${AVIF_LIBRARIES}` isn't set correctly.

cryptomilk replied:
There is no `AVIF_LIBRARIES` variable, `avif` is a target!

vapier replied:
sorry, but i don't know what that means.  we should not be hardcoding `avif` as a library name.  the entire point of using pkg-config (or a cmake module) is for it to detect the right linker settings and set up a variable (like `AVIF_LIBRARIES`) for the rest of the files to use.

cryptomilk replied:
Please take a look at `/usr/lib64/cmake/libavif/libavif-config-*.cmake`

The avif TARGET gets imported in the config see
`set_target_properties(avif PROPERTIES ...)`

Details on target properties are documented here:
https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html#target-properties

vapier replied:
as long as the libavif cmake module declares `avif` such that it is expanded into the correct library for `target_link_libraries` (i.e. "A library target name"), this is fine.  but we cannot have `target_link_libraries` treat `avif` as `-lavif` using `target_link_libraries`'s own resolution logic (i.e. "A plain library name").

vapier replied:
i took a look at this locally.  it looks like `avif` is treated as a "a library target name" and so expands into the cmake settings.  however, it shows how fundamentally broken and/or poorly designed cmake is: it loads the full path to the libavif library and links against that directly.  this breaks fundamental cross-compiling principles.

that said, if anyone wants to use cmake, they're opting into the broken cmake design.  the autotools code cross-compiles just fine.

@morsssss you can delete these lines and change the toplevel CMakeLists.txt instead:

```
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -123,6 +123,8 @@ else (USE_EXT_GD)
    IF (ENABLE_AVIF)
        FIND_PACKAGE(libavif 0.8.2 REQUIRED CONFIG)
        SET(HAVE_LIBAVIF 1)
+       SET(AVIF_LIBRARIES avif)
+       SET(AVIF_FOUND 1)
    ENDIF (ENABLE_AVIF)
 
    IF (ENABLE_LIQ)
```

this builds off of the target logic that @cryptomilk wants but keeps the existing gd cmake style so we can merge this without blocking on rewriting the cmake files to use the target logic directly.

morsssss replied:
Got it. This worked for me as well - moving all the logic to the top-level file, and letting the subsidiary `CMakeLists.txt` files use our usual flags for logic instead of targets.

vapier commented on src/gd.h:
```diff
@@ -627,6 +627,10 @@ BGD_DECLARE(gdImagePtr) gdImageCreateFromWebp (FILE * inFile);
 BGD_DECLARE(gdImagePtr) gdImageCreateFromWebpPtr (int size, void *data);
 BGD_DECLARE(gdImagePtr) gdImageCreateFromWebpCtx (gdIOCtx * infile);
 
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvif (FILE *inFile);
```
omit the space before the `(`

vapier commented on src/gd.h:
```diff
@@ -1084,6 +1088,11 @@ BGD_DECLARE(void *) gdImageWebpPtr (gdImagePtr im, int *size);
 BGD_DECLARE(void *) gdImageWebpPtrEx (gdImagePtr im, int *size, int quantization);
 BGD_DECLARE(void) gdImageWebpCtx (gdImagePtr im, gdIOCtx * outfile, int quantization);
 
+BGD_DECLARE(void) gdImageAvif (gdImagePtr im, FILE *outFile);
```
omit the space before the `(`

vapier commented on tests/avif/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/jpeg2gd
```
you'll need to update this file

vapier commented on tests/avif/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+IF(TARGET avif)
```
this looks incorrect.  we write `IF(LIBAVIF_FOUND)` in files instead (like you see below w/png & jpeg).

cryptomilk replied:
This looks absolutely correct. You should learn modern cmake, read up on targets and interfaces.

https://cliutils.gitlab.io/modern-cmake/chapters/intro/dodonot.html

Several of the "Don'ts" apply to libgd.

vapier replied:
cmake is not a well supported build system in gd.  our current style is based on FOUND vars, not TARGET.  if we want to rewrite things to use a new pattern, especially if it's one cmake recommends, we can do that, but we should do it across all the build files.  people are just copying & pasting what's already here when writing new code.

keep in mind that we currently support cmake-3.7+.  so any features that require newer versions won't work.

we could prob bump it up to cmake-3.10, but not further at this point (going by Ubuntu Bionic being 3.10).

ftr, i have no real interest in cmake.  i'd just as soon delete it and be done.

vapier commented on tests/avif/Makemodule.am:
```diff
@@ -0,0 +1,23 @@
+	avif/avif_null
+
+if HAVE_LIBPNG
+	avif/png2avif
```
missing `libgd_test_programs += \` line

morsssss replied:
thank you

vapier reviewed (changes requested):
i think it's fine to emit a warning for now via `gd_error(...)`.  we already expose a framework to users with `gdSetErrorMethod` so they can capture/display all such messages.

move your little test examples to the `examples/` dir and call it a day.  we don't build any of those atm, and we should fix that, but we don't have to hold this PR up for it.

wrt the tests, i wouldn't worry about 100% coverage ... just do a basic set that you think is reasonable.  but the test files seem to be larger than necessary.  smaller is better here as the git history is forever.  so shrink them down -- ideally <50k per file.

i'll take a look at the generated cmake header file.

vapier commented:
travis failures should be fixed now.  i broke it earlier due to not fully reading/understanding the logs.  sorry about that.

you'll need to rebase onto the latest master.

morsssss commented:
> travis failures should be fixed now. i broke it earlier due to not fully reading/understanding the logs. sorry about that.
> 
> you'll need to rebase onto the latest master.

Thanks!

Sadly when I tried to rebase, git kept getting stuck resurrecting old commits and asking me to resolve conflicts there. Finally I just did a merge. Sorry about the redundant commit logs.

vapier commented on CMakeLists.txt:
```diff
@@ -29,27 +30,27 @@ endif(BUILD_TEST)
 # - specify CFLAGS=... on invocation of make
 IF(CMAKE_COMPILER_IS_GNUCC)
-  SET(CMAKE_C_FLAGS_DEBUG "-g -Wall -Wextra -O0") # will be added to CMAKE_C_FLAGS when CMAKE_BUILD_TYPE is "Debug"
+	SET(CMAKE_C_FLAGS_DEBUG "-g -Wall -Wextra -O0") # will be added to CMAKE_C_FLAGS when CMAKE_BUILD_TYPE is "Debug"
```
looks like you've reformatted unrelated lines.  should omit these.

morsssss replied:
Yes, I got inspired, but I can remove them.

vapier commented on examples/jpeg2avif.c:
```diff
@@ -0,0 +1,57 @@
+
+	gdImageAvif(im, out);
+
+  printf("Wrote outfile %s.\n", argv[2]);
```
use tabs

vapier commented on examples/jpeg2avifex.c:
```diff
@@ -0,0 +1,99 @@
+		usage();
+
+	infile = malloc((strlen(argv[optind]) + 1) * sizeof(char));
+	strcpy(infile, argv[optind++]);
```
just use `strdup()`

morsssss replied:
I should have known there was an easier way

vapier commented on examples/jpeg2avifex.c:
```diff
@@ -0,0 +1,99 @@
+#include "gd.h"
+
+#ifdef HAVE_LIBAVIF
+#include <avif/avif.h>
```
do you need this include ?  doesn't seem like it.

morsssss replied:
I guess I don't!

morsssss replied:
argh

vapier commented on src/CMakeLists.txt:
```diff
@@ -134,11 +135,14 @@ SET(LIBGD_DEP_LIBS
 if (BUILD_SHARED_LIBS)
 	target_link_libraries(${GD_LIB} ${LIBGD_DEP_LIBS})
 endif()
+
```
omit these new blank lines

vapier commented on src/gd.h:
```diff
@@ -1084,6 +1088,11 @@ BGD_DECLARE(void *) gdImageWebpPtr (gdImagePtr im, int *size);
+BGD_DECLARE(void) gdImageAvifEx(gdImagePtr im, FILE *outFile, int quality, int speed);
+BGD_DECLARE(void *) gdImageAvifPtr(gdImagePtr im, int *size);
+BGD_DECLARE(void *) gdImageAvifPtrEx(gdImagePtr im, int *size, int quality, int speed);
+BGD_DECLARE(void) gdImageAvifCtx(gdImagePtr im, gdIOCtx * outfile, int quality, int speed);
```
no space after the `*`

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,684 @@
+	io->write = NULL; // this function is currently unused; see avif.h
+	io->destroy = destroyAvifIO;
+	io->sizeHint = 0; // sadly, we don't get this information from the gdIOCtx.
+	io->persistent = AVIF_FALSE; // TODO: This seems like the safe thing to do, but I don't know for sure. This is less efficient, because it means AVIF will make copies of all buffers.
```
should avoid inline comments like this.  they're very long and hard to read.  move it above the line and line wrap it.

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,684 @@
+	}
+
+	cleanup:
+		avifDecoderDestroy(decoder);		// if io has been allocated, this frees it
```
don't indent the goto targets

vapier replied:
still needs fixing

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,684 @@
+
+	if (!gdImageSX(im) || !gdImageSY(im)) {
+		gd_error("at least one image dimension is zero");
+		return(1);
```
`return` isn't a function

morsssss commented:
I've resolved the remaining issues. I made a few more changes recommended by @wantehchang . And I added two more little test suites.

* `bad_input.c` tries the AVIF functions with a few bad parameters. This passes.
* `compare_avif_to_png.c` works with a set of reference PNG and AVIF files. It transforms each PNG to an AVIF and compares each with the original PNG. Then it does the reverse - it transforms each AVIF to a PNG and compares each with the original AVIF.

Unfortunately, the latter test works in only one direction. Encoding PNGs to AVIFs, which is where I'd expect we'd have trouble, works great. Decoding AVIFs and re-encoding the result as a PNG results in minor discrepancies - little differences in almost every pixel.

I'm a bit at a loss here.
* `im2_im` passes - so we've achieved pixel perfection with encoding/decoding simple images, all within the GD system.
* Unless I'm wrong, **encoding** is where we could experience some lossiness, not **decoding**. So why do we have trouble when decoding an AVIF to the gd format, decoding a reference PNG to the gd format, and comparing the two?

In fact, it turns out that my AVIFs and my reference PNGs are not quite the same - at least, they look subtly different in Chrome. Worse yet, if I encode a PNG to AVIF in squoosh.app, the results are a bit different. Even the PNGs and AVIFs [here](https://github.com/link-u/avif-sample-images) look different.

I realize as well that other image types don't have tests like this.

Am I missing something obvious here?

morsssss commented:
Finally, I just added `-DENABLE_AVIF=1` to the `cmake` run by Travis. It's not able to find `libavif`, presumably because it's not installed. Is this the same issue that was encountered in https://github.com/libgd/libgd/pull/672 ?

vapier commented:
the travis build failure isn't the same.  the build env doesn't include libavif by default.  it's using Ubuntu Xenial which ships with libavif-0.7 which i'm guessing is too old.  but you can try to update `.travis.yml` to list the package and see if it builds.

vapier commented:
i see the cmake file explicitly requires libavif-0.8.2+, so we won't be able to test it in travis for a while.  glancing through [their docs](https://docs.travis-ci.com/user/reference/linux), Ubuntu Focal is the most recent version they support, but even that still only has [libavif-0.7](https://packages.ubuntu.com/search?keywords=libavif&searchon=names&suite=focal&section=all).

seems Ubuntu hasn't upgraded to newer libavif until [hirsute](https://packages.ubuntu.com/search?keywords=libavif&searchon=names&suite=hirsute&section=all) and there it has libavif-0.8.4.

vapier commented on configure.ac:
```diff
@@ -285,6 +285,14 @@ GD_LIB_PKG_CHECK([LIBWEBP], [WEBP], [webp], [libwebp], [
 ])
 
+dnl Check for avif support.
+GD_LIB_PKG_CHECK([LIBAVIF], [AVIF], [avif], [libavif], [
```
since there's a min version requirement, you'll want to express that here too

`GD_LIB_PKG_CHECK([LIBAVIF], [AVIF], [avif], [libavif >= 0.8.2], [`

vapier commented on tests/avif/compare_avif_to_png.c:
```diff
@@ -0,0 +1,93 @@
+	void *avifImDataPtr = NULL, *pngImDataPtr = NULL;
+	int size;
+	char filePath[MAX_FILEPATH_LENGTH], pngFilePath[MAX_FILEPATH_LENGTH], avifFilePath[MAX_FILEPATH_LENGTH];
+  char errMsg[MAX_FILEPATH_LENGTH + 100];
```
the whitespace in this file is all funky

morsssss replied:
I keep grepping for bad whitespace, and yet somehow I sometimes miss some 😭

morsssss commented:
> i see the cmake file explicitly requires libavif-0.8.2+, so we won't be able to test it in travis for a while. glancing through [their docs](https://docs.travis-ci.com/user/reference/linux), Ubuntu Focal is the most recent version they support, but even that still only has [libavif-0.7](https://packages.ubuntu.com/search?keywords=libavif&searchon=names&suite=focal&section=all).
> 
> seems Ubuntu hasn't upgraded to newer libavif until [hirsute](https://packages.ubuntu.com/search?keywords=libavif&searchon=names&suite=hirsute&section=all) and there it has libavif-0.8.4.

We chose v0.8.2 simply because it was a recent version, not because we particularly need 0.8.2. Does this mean we're now keeping out distributions that have older versions of libavif? If so, can we not require 0.8.2?

vapier commented:
> If so, can we not require 0.8.2?

i have no idea.  if the API is compatible between 0.7 & 0.8.2, i don't see a problem lowering the requirement.

cryptomilk commented:
> If the API is compatible between 0.7 & 0.8.2, i don't see a problem lowering the requirement.

As I said, the API changed between versions. If you want to support more than one API you will start to introduce `#ifdef AVIF_VERSION` and have to implement probably 3 different APIs.

morsssss commented:
> > If the API is compatible between 0.7 & 0.8.2, i don't see a problem lowering the requirement.
> 
> As I said, the API changed between versions. If you want to support more than one API you will start to introduce `#ifdef AVIF_VERSION` and have to implement probably 3 different APIs.

Thanks for this, cryptomilk.

@wantehchang , do you see any easy path toward using v0.7 instead?
Of course I'd hate to re-do this work - and I'd imagine there are many reasons to use a newer version of libavif. 
On the other hand, [it does look like](https://packages.ubuntu.com/search?keywords=libavif) no released version of Ubuntu includes v0.8... not until [hirsute is released in April](https://discourse.ubuntu.com/t/hirsute-hippo-release-schedule/18539).

I don't know about other distros.

vapier commented:
lets ship this as-is (requiring 0.8.4+) and re-evaluate based on user feedback.  most people are getting gd via their distros rather than installing it themselves, so it's not like the new gd is going to be shipped on older distros either.

morsssss commented:
> most people are getting gd via their distros rather than installing it themselves, so it's not like the new gd is going to be shipped on older distros either.

That... is a really good point.

Let us proceed!

morsssss commented:
@vapier, I'd be inclined to take AVIF back out of the Travis build... wdyt?

vapier commented:
yes, it cannot be enabled in Travis if it's not going to pass ;)

you can comment out the line with a TODO entry saying "When we switch to Ubuntu 21+ (Hirsute), we can reenable libavif coverage.".

morsssss commented:
> you can comment out the line with a TODO entry saying "When we switch to Ubuntu 21+ (Hirsute), we can reenable libavif coverage.".

Will do!

Meanwhile, about the image discrepancies I'm experiencing... I finally tried decoding a PNG and simply re-encoding it as a PNG, like this:

```c
im = gdImageCreateFromPng(in);
gdImagePng(im, out);
```

I saw the same discrepancies. As in, the AVIF image I generated like this:

```c
gdImageAvifEx(im, out, 100, 0);
```

was the same as the re-encoded PNG.

I then tried the same thing on a basic PNG found [here](http://www.schaik.com/pngsuite/pngsuite_bas_png.html). It, too, looked different when decoded and re-encoded.

Any ideas here would be helpful...

Quy commented on travis/main.sh:
```diff
@@ -94,7 +97,7 @@ cmake_args=(
 )
 
 # libxpm-dev is unavaible in brew repo
```
```suggestion
# libxpm-dev is available in brew repo
```

vapier replied:
while "unavaible" is misspelled, it should not be changed to "available" here

Quy replied:
Oops. Sorry copy and paste error.

morsssss replied:
perhaps a nice case for a little separate PR?

morsssss commented:
I've tried encoding to PNG's with different options and different PNG's - but I still get differences in the colors. I feel like I'm not the first person to have trouble with this sort of test - noticing the test in `jpeg_im2im.c` that's `#ifdef`'d out.

The tests confirm that encoding and decoding AVIFs happen losslessly - so for now I've commented out the offending test in `compare_avif_to_png.c`. I don't love this solution, but I also don't want to hold up this PR for a problem that really doesn't seem to lie in the AVIF encoding or decoding.

morsssss commented:
I've squashed all the commits down to a much smaller number, and removed the merges with the main repo... is this good to go?

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,677 @@
+	void *dataBuf = NULL;
+	gdIOCtx *ctx = (gdIOCtx *) io->data;
+
+	//TODO: if we set sizeHint, this will be more efficient.
```
space after `//`

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,677 @@
+	 Returns NULL if memory for the object can't be allocated.
+*/
+
+//TODO: can we get sizeHint somehow?
```
space after the `//`

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,677 @@
+	if (io == NULL)
+		return NULL;
+
+// TODO: setting persistent=FALSE is safe, but it's less efficient. Is it necessary?
```
indent it

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,677 @@
+
+	im->saveAlphaFlag = 1;
+
+// Read the pixels from the AVIF image and copy them into the GD image.
```
indent it

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,677 @@
+	avifIm->matrixCoefficients = lossless ? AVIF_MATRIX_COEFFICIENTS_IDENTITY : AVIF_MATRIX_COEFFICIENTS_BT709;
+
+	avifRGBImageSetDefaults(&rgb, avifIm);
+	avifRGBImageAllocatePixels(&rgb); // this allocates memory, and sets rgb.rowBytes and rgb.pixels.
```
avoid inline comments

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,677 @@
+
+	encoder = avifEncoderCreate();
+	int quantizerQuality = quality == QUALITY_DEFAULT ?
+												 QUANTIZER_DEFAULT : quality2Quantizer(quality);
```
bad indentation

morsssss replied:
is this better? I just indented the second line by a single tab, instead of trying to line up the second line horizontally with the word "quality".

morsssss replied:
And I just made the same change a few lines earlier:

```c
	avifPixelFormat subsampling = quality >= HIGH_QUALITY_SUBSAMPLING_THRESHOLD ?
		CHROMA_SUBAMPLING_HIGH_QUALITY : CHROMA_SUBSAMPLING_DEFAULT;
```

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,677 @@
+	if (failed)
+		goto cleanup;
+
+	result = avifEncoderAddImage(encoder, avifIm, 1, AVIF_ADD_IMAGE_FLAG_SINGLE); //TODO: is there a reason to use timeSscales != 1?
```
uninline the comment

vapier commented on src/gd_avif.c:
```diff
@@ -0,0 +1,677 @@
+	gdPutBuf(avifOutput.data, avifOutput.size, outfile);
+
+cleanup:
+		if (rgb.pixels)
```
unindent the goto blocks

vapier reviewed (changes requested):
mostly nits

@cryptomilk want to take one last pass before we merge ?

morsssss commented:
Ok, I've rebased this on the new master after #670 was merged.

Merging, I realize AVIF support was missing from `windows/Makefile.vc` and `windows/msys/Makefile` in a couple of places where HEIF support had been added. I've added that.

I combined the new `README.md` comments... lmk if you want that changed.

I also see that the `FType enum` was removed from `gd_filename.c` in [this commit](https://github.com/libgd/libgd/commit/f538eedc7cf1dbf3d84148432b0fe61e5b3789a0#diff-56e804fadcc3fc36abc8dca2fb5fe50d90d09abd7ebe1d776e563eeb9c3ffad6). I didn't have that before this rebase, but I do now.

I still haven't tested the merged code...

vapier commented on CMakeLists.txt:
```diff
@@ -120,9 +121,12 @@ else (USE_EXT_GD)
 		FIND_PACKAGE(WEBP REQUIRED)
 	ENDIF (ENABLE_WEBP)
 
-	IF (ENABLE_HEIF)
```
don't delete the code we just added

morsssss replied:
thanks for spotting that!

vapier commented on README.md:
```diff
@@ -40,7 +40,8 @@ GD has builtin support for:
 * [HEIF](https://en.wikipedia.org/wiki/High_Efficiency_Image_File_Format) via [libheif](https://github.com/strukturag/libheif/)
-  * Also it includes [AVIF](https://en.wikipedia.org/wiki/AV1#AV1_Image_File_Format_%28AVIF%29) read support if your system's `libheif` has AV1 decoding.
+  * This includes [AVIF](https://en.wikipedia.org/wiki/AV1#AV1_Image_File_Format_%28AVIF%29) read support if your system's `libheif` has AV1 decoding.
+* [AVIF](https://en.wikipedia.org/wiki/AV1#AV1_Image_File_Format_(AVIF)) via [libavif](https://github.com/AOMediaCodec/libavif)
```
"A" before "H"

morsssss replied:
@#$%@#^&@# merge

vapier commented on examples/jpeg2avifex.c:
```diff
@@ -0,0 +1,102 @@
+
+#include "gd.h"
+
+#ifdef HAVE_LIBAVIF
```
don't want these ifdef checks
